src/receiver.rs

```diff
@@ -1,78 +1,344 @@
+            {
+                // Without async, we just set the flag and check we won.
+                let flags = hatch.flags.fetch_or(R_CLOSE, orderings::MODIFY);
+                if any_flag(flags, S_CLOSE) {
+                    // Safe because we have seen the sender closed
+                    unsafe { hatch.cleanup(any_flag(self.flags, MARK_ON_DROP)); }
+                }
+            }
+            #[cfg(feature="async")] {
+                // For async, we have to lock and maybe wake the sender
+                let flags = hatch.lock();
+                if any_flag(flags, S_CLOSE) {
+                    // Safe because we have seen the sender closed
+                    return unsafe { hatch.cleanup(any_flag(self.flags, MARK_ON_DROP)); }
+                } else {
+                    let shared = unsafe { &mut *hatch.inner.get() };
+                    let _recv = shared.receiver.take(); // we won't be waiting any more
+                    let sender = shared.sender.take();  // we might need to wake them
+                    // mark us closed and release the lock in one smooth action.
+                    hatch.flags.store(flags | R_CLOSE, orderings::STORE);
+                    // Finally, wake the sender if they are waiting.
+                    if let Some(waker) = sender { waker.wake(); }
+                }
+            }
         }
-        let state = this.inner.set_recv(ctx.waker().clone());
-        match this.handle_state(state) {
-            Poll::Pending => {},
-            x => return x,
+    }
+}
+
+/// This is a disposable object which performs a single receive operation.
+///
+/// The `now` method attempts synchronously to receive a value, but
+/// there may not be a value in the hatch.
+///
+/// ## Options (all false by default)
+///
+/// * `close_on_receive` - when true, closes when we successfully receive a message.
+/// * `mark_on_drop` - when true, cleanup marks the atomic with a sentinel value for an external
+///   process (not provided!) to reclaim it.
+///
+/// Notes:
+/// * Setting `close_on_receive` on this object will only affect this object.
+/// * `mark_on_drop` must be set on the `Receiver`.
+///
+/// ## Async support (feature `async`, default enabled)
+///
+/// This struct is a `Future` which polls ready when either the
+/// receive was successful or the `Sender` has closed.
+#[derive(Debug)]
+pub struct Receiving<'a, 'b, T> {
+    receiver: Option<&'b mut Receiver<'a, T>>,
+    flags: Flags,
+}
+
+impl<'a, 'b, T> Receiving<'a, 'b, T> {
+    /// Gets the value of the `close_on_receive` option. the [`Receiving`] docs for an explanation.
+    pub fn get_close_on_receive(&self) -> bool { any_flag(self.flags, CLOSE_ON_SUCCESS) }
+
+    /// Sets the `close_on_receive` option. See the [`Receiving`] docs for an explanation.
+    pub fn close_on_receive(mut self, on: bool) -> Self {
+        self.flags = toggle_flag(self.flags, CLOSE_ON_SUCCESS, on);
+        self
+    }
+}
+
+#[cfg(not(feature="async"))]
```

Comment:
I think it would be a better idea to use `cfg_attr` on the doc comment, and `#[cfg(...)]` in a more fine-grained manner, because it might be more difficult to keep/maintain these two versions otherwise.

Reply:
Yeah, fwiw i hate both options :)